syntax = "proto2";

import public "method_options.proto";
import public "service_options.proto";
import public "field_options.proto";
package Battlenet;

option optimize_for = CODE_SIZE;

message NO_RESPONSE {
}

message Address {
  required string address = 1;
  optional uint32 port = 2;
}

message ProcessId {
  required uint32 label = 1;
  required uint32 epoch = 2;
}

message ObjectAddress {
  required .Battlenet.ProcessId host = 1;
  optional uint64 object_id = 2 [default = 0];
}

message NoData {
}

message ErrorInfo {
  required .Battlenet.ObjectAddress object_address = 1;
  required uint32 status = 2;
  required uint32 service_hash = 3;
  required uint32 method_id = 4;
}

message Header {
  required uint32 service_id = 1;
  optional uint32 method_id = 2;
  required uint32 token = 3;
  optional uint64 object_id = 4 [default = 0];
  optional uint32 size = 5 [default = 0];
  optional uint32 status = 6 [default = 0];
  repeated .Battlenet.ErrorInfo error = 7;
  optional uint64 timeout = 8;
  optional bool is_response = 9;
  repeated .Battlenet.ProcessId forward_targets = 10;
  optional fixed32 service_hash = 11;
}


